### PR TITLE
msg: fixup for 2ffacbe (crc configuration in messenger)

### PR DESCRIPTION
**Pull request by trociny**
* Update the remaining Message::encode() calls, which now expect crc
  flags to be passed as a flags parameter instead of a bool (this e.g.
  fixes the issue with routing messages forwarded from an older client).

* In Message::encode() data crc is calculated when MSG_CRC_DATA is
  set, but in decode_message(), Pipe::read/write_message() data crc is
  calculated when MSG_CRC_HEADER is set. Fix this.

Note, 2ffacbe changed the behavior of Pipe::read_message/write_message():
previously the method always calculated crc, now they calculate it
only if crc is enabled in the config. This means crc can not be disabled
if there are monitors of older version in the cluster.

Signed-off-by: Mykola Golub <mgolub@mirantis.com>

### CODE REVIEW

**Comment by yuyuyu101** [comment]
This commit reminds me that the compatible of the crc code changes. If one is older version which encode message, and server side will decode message according to config value. It seemed affect existing/older clients? We may need to get crc flags from message not from config.

@liewegas

File: src/msg/simple/Pipe.cc [hunk 1]
```diff
@@ -2035,8 +2034,9 @@ int Pipe::read_message(Message **pm, AuthSessionHandler* auth_handler)
     if (msgr->crcflags & MSG_CRC_HEADER) {
       footer.front_crc = old_footer.front_crc;
       footer.middle_crc = old_footer.middle_crc;
-      footer.data_crc = old_footer.data_crc;
     }
+    if (msgr->crcflags & MSG_CRC_DATA)
+      footer.data_crc = old_footer.data_crc;
     footer.sig = 0;
```
Comment (liewegas) [inline src/msg/simple/Pipe.cc]: i think this block may as well unconditionally copy these fields.. we're just translating the footer formats.  the checks some later and we may ignore them but that's fine

**Comment by liewegas** [comment]
I wonder if we should also add (as an additional patch) a new footer flag for no header crc, similar to MSG_FOOTER_NOCRC (which we could rename NODATACRC)?  that way we know what to check or not check based on what the sender did.. does that resolve all of the compat issues?

**Comment by yuyuyu101** [comment]
Yeah, and what's more I have a another need that we want to separate calculate header crc and others. Because AsyncMessenger want to calculate header crc at last before we will send it. So AsyncMessenger can avoid lock conflicting when multi thread call send_message.

**Comment by trociny** [comment]
Rebased after applying liewegas' suggestion about copying unconditionally footer crc fields.

For an additional footer flag, it might be useful for future, but I
don't see how it can help with the compatibility issue I am aware of.
The issue is that the old Pipe::read_message() unconditionally checks
the header crc.  The old Pipe::write_message() unconditionally sends
crc, so communication between old mons with crc off works. But the new
Pipe::write_message() calculates and sends the header crc only if
ms_crc_header is on.  As a result, if crc is disabled in the cluster
with old and new mons (ms_nocrc = yes, for old, and ms_crc_header =
ms_crc_data = false for new), errors like below are observed on the old  mon:

 2015-01-23 11:20:37.078114 7fa48df8b700  0 -- 172.18.128.29:6791/0 >> 172.18.128.29:6790/0 pipe(0x41a8000 sd=24 :6791 s=2 pgs=27503 cs=54987 l=0 c=0x3b16420).reader got bad header crc 63074668 != 0

A workaround is to enable crc until all monitors are updated. It may
be fixed by making Pipe::write_message() unconditionally calculate the
header crc again. Or may be to detect somehow the old mon presence and
calculate crc only in this case.  But is this worth doing? I don't
expect to see many such configurations in production, while the
workaround exists. Still it is good to document.

**Comment by trociny** [comment]
Also added CEPH_MSG_FOOTER_NOHEADERCRC flag as it was suggested.

I think it would be also nice to have CEPH_MSG_HEADER_NOCRC, but we don't have header.flags field for this.

**Comment by yuyuyu101** [comment]
Hmm, I would like to clarify more. If one side disable header crc and another side enable header crc, we need to check message's footer flag to judge whether we need to compare(This check is happened in pipe.cc and asyncconnection.cc). It will result in that we need to delay to check header crc until we receive all message content. But it's not reasonable that we can read enough network buffer if header crc is wrong, it looks like that we need to add flags to header?

**Comment by mattbenjamin** [comment]
The detailed checks weren't really intended to be used to get this fine-grained control message-by-message.  For that, something more transactional/negotiated between a pair of endpoints might work better?  Just thinking aloud...

**Comment by yuyuyu101** [comment]
Hmm, it maybe better exchange crc settings in connection handshake period and set flag in ceph_msg_connect and ceph_msg_connect_reply

**Comment by trociny** [comment]
Added yet another improvement: in Pipe::write_message, when copying crc to old footer, use CEPH_MSG_FOOTER_NOHEADERCRC/NODATACRC flags instead of config options.

So, from the discussion, to improve detection if crc is needed, it looks we have two routes to go:

i) Add header.flags and CEPH_MSG_FOOTER_NOCRC flag to messages. Only check header crc for messages with this flag set.
ii) Detect if crc is enabled on a connection handshake and use this result when calculating/checking message crc.

I ready to work on any of the solutions if it is agreed what route to choose. I prefer (ii) because

* we need to detect only once per connection
* if any side disabled crc, both sides would not do calculation (In (i) if one side has crc enabled it still sends messages with crc calculated).

Also, I suppose we need to sync with the kernel client. I think we already have some problems that the kernel client still uses old option nocrc.

That said, I think this is out of the scope of the initial request to fix the issue with routing messages forwarded from an older client. That looks most critical and currently is a blocker for my other work -- disable forwarding for some messages.
So it would be nice to merge at least the first commit (the rest might be not necessary if choose the route (ii)).

**Comment by liewegas** [comment]
I agree that setting these flags during the handshake (CEPH_MSG_CONNECT_NOHEADERCRC/DATACRC flags) makes the most sense.

**Comment by trociny** [comment]
@liewegas, do you expect me to work on setting crc flags during the handshake (CEPH_MSG_CONNECT_NOHEADERCRC/DATACRC flags)?

I saw you merged my patches to  #3490 so thought some general solution was going to be implemented there. Still if you think it is helpful to implement CEPH_MSG_CONNECT_NOHEADERCRC/DATACRC flags in this branch I will be glad to work on it.

Actually, I do like to see the first patch from this pull (d5d1e27) merged to master ASAP, as it fixes forwarding broken between new and old monitors.

**Comment by liewegas** [comment]
merged the first patch manually to master.  i think you're right that the footer flag for the header crc isn't terribly helpful, and having per-message granularity isn't useful.

the per session negotiation seems like the way to go, but to be honest unless this is bugging you i think there are more important things to look at..

**Pull request closed without merging**